runsc/container/container.go

```diff
@@ -912,19 +927,28 @@ func (c *Container) forEachSelfMount(fn func(mountSrc string)) {
 // createGoferFilestores creates the regular files that will back the
 // tmpfs/overlayfs mounts that will overlay some gofer mounts. It also returns
 // information about how each gofer mount is configured.
-func (c *Container) createGoferFilestores(ovlConf config.Overlay2, mountHints *boot.PodMountHints) ([]*os.File, []boot.GoferMountConf, error) {
+func (c *Container) createGoferFilestores(ovlConf config.Overlay2, mountHints *boot.PodMountHints, rootfsHint *boot.RootfsHint) ([]*os.File, []boot.GoferMountConf, error) {
 	var goferFilestores []*os.File
 	var goferConfs []boot.GoferMountConf
 
 	// Handle root mount first.
-	shouldOverlay := ovlConf.RootEnabled() && !c.Spec.Root.Readonly
-	filestore, goferConf, err := c.createGoferFilestore(ovlConf, c.Spec.Root.Path, shouldOverlay, nil /* hint */)
+	overlayMedium, shouldOverlay := ovlConf.Medium(), ovlConf.RootEnabled()
+	if rootfsHint != nil && rootfsHint.Overlay != config.NoOverlay {
+		overlayMedium, shouldOverlay = rootfsHint.Overlay, true
+	}
+	shouldOverlay = shouldOverlay && !c.Spec.Root.Readonly
+	filestore, goferConf, err := c.createGoferFilestore(overlayMedium, c.Spec.Root.Path, shouldOverlay, nil /* hint */)
 	if err != nil {
 		return nil, nil, err
 	}
 	if filestore != nil {
 		goferFilestores = append(goferFilestores, filestore)
 	}
+	if rootfsHint != nil && rootfsHint.Mount.Type != boot.Bind {
+		if err := goferConf.Lower.Set(rootfsHint.Mount.Type); err != nil {
+			return nil, nil, fmt.Errorf("failed to override root mount lower layer config: %v", err)
+		}
+	}
```

Comment:
Agree! And thanks for the code! I have the same concerns. I tried something similar and failed to make the code as short as the old one. I wasn't sure if it's what we want. So I just gave it up before finishing it.

I just pushed the updated version. In this version, I also changed the string representation of `config.NoOverlay` from `""` to `"none"`. Because, `"dev.gvisor.spec.rootfs.overlay": "none"` looks more digestible than `"dev.gvisor.spec.rootfs.overlay": ""` when representing the meaning of disabling overlay. And it will also allow us to inherit the overlay config set by the runtime flags by omitting the overlay annotation.

PS. I replaced the "Suggested-by:" line with a "Co-authored-by:" line in [this patch](https://github.com/google/gvisor/pull/9486/commits/0126e204904289564c6ed0e1999bda92e2f0c61b). Please let me know, if you don't want this. Thanks again!

Reply:
Left a comment about the `none` medium value.

> And it will also allow us to inherit the overlay config set by the runtime flags by omitting the overlay annotation.
I think this can be confusing. Sub mount hints don't interplay at all with `--overlay2` flag. I think these should be kept separate with no overlap.

Thanks for adding me as co-author :)